asm/vm/asm_vm

```diff
@@ -648,11 +795,68 @@ validate_asm_installation() {
     fatal "ASM ${ASM_RELEASE} version is not found. Please install ASM ${ASM_RELEASE} and retry."
   fi
 
-  if [[ "${EXPANSION_GATEWAY_INSTALLED}" -eq 0 ]]; then
+  if [[ "${AUTOREGISTRATION_ENABLED}" -eq 0 ]]; then
     { read -r -d '' MSG; fatal "${MSG}"; } <<EOF || true
+VM auto-registration is not enabled in any of your ASM installation in the
+cluster. Please use install_asm script to install and make sure none of the
+values set in your custom overlay, if any, overrides the default values.
+EOF
+  fi
+
+  if [[ "${ENVIRON_WORKLOAD_IDENTITY_USED}" -eq 0 ]]; then
+    { read -r -d '' MSG; fatal "${MSG}"; } <<EOF || true
+The Environ workload identity is not used in any of your ASM installation in the
+cluster. Please make sure "--option hub-meshca" is included when you
+use the install_asm script.
+EOF
+  fi
+
+  if [[ "${EXPANSION_GATEWAY_INSTALLED}" -eq 0 ]]; then
+    if [[ "${CREATE_GCE_INSTANCE_TEMPLATE}" -eq 1 ]] || [[ "${ONLY_VALIDATE}" -eq 1 ]]; then
+      { read -r -d '' MSG; fatal "${MSG}"; } <<EOF || true
 ${EXPANSION_GATEWAY_NAME} is not found in the cluster.
-Please use install_asm with "--option vm" when you install ASM.
+Please use install_asm script with "--option vm" when you install ASM or run
+the current script with the "prepare_cluster" subcommand.
 EOF
+    fi
+    if [[ "${PREPARE_CLUSTER}" -eq 1 ]]; then
+      INSTALL_EXPANSION_GATEWAY=1
+    fi
+  fi
+}
+
+validate_google_identity_provider() {
+  info "Verifying identity providers in the cluster..."
+  if ! is_google_identity_provider_installed; then
+    if [[ "${CREATE_GCE_INSTANCE_TEMPLATE}" -eq 1 ]] || [[ "${ONLY_VALIDATE}" -eq 1 ]]; then
+      { read -r -d '' MSG; fatal "${MSG}"; } <<EOF || true
+GCE identity provider is not found in the cluster. Please run the install_asm
+script with "--option vm" flag to allow the script to register the google
+identity provider in your cluster or run "prepare_cluster" subcommand of this
+script.
+EOF
+    fi
+    if [[ "${PREPARE_CLUSTER}" -eq 1 ]]; then
+      INSTALL_IDENTITY_PROVIDER=1
+    fi
+  fi
+}
+
+is_google_identity_provider_installed() {
+  if ! is_idp_crd_installed; then
+    false
+    return
+  fi
+
+  if [[ "$(retry 2 kubectl get identityproviders.security.cloud.google.com -ojsonpath="{..metadata.name}" \
+    | grep -w -c google || true)" -eq 0 ]]; then
```

Comment:
style nit: This should work but I'm not 100% sure with the retry there.
```
if ! retry 2 kubectl get identityproviders.security.cloud.google.com -ojsonpath="{..metadata.name}" \
    | grep -w -q google; then
```

Reply:
done.